socs/agents/scpi_psu/agent.py

```diff
@@ -68,28 +86,38 @@ def monitor_output(self, session, params=None):
                 Defaults to False.
 
         """
-        session.set_status('running')
         self.monitor = True
 
         while self.monitor:
             with self.lock.acquire_timeout(1) as acquired:
                 if acquired:
-                    data = {
-                        'timestamp': time.time(),
-                        'block_name': 'output',
-                        'data': {}
-                    }
-
-                    for chan in params['channels']:
-                        data['data']["Voltage_{}".format(chan)] = self.psu.get_volt(chan)
-                        data['data']["Current_{}".format(chan)] = self.psu.get_curr(chan)
-
-                    # self.log.info(str(data))
-                    # print(data)
-                    self.agent.publish_to_feed('psu_output', data)
-
-                    # Allow this process to be queried to return current data
-                    session.data = data
+                    if self.psu:
+                        data = {
+                            'timestamp': time.time(),
+                            'block_name': 'output',
+                            'data': {}
+                        }
+
+                        for chan in params['channels']:
+                            try:
+                                data['data']["Voltage_{}".format(chan)] = self.psu.get_volt(chan)
+                                data['data']["Current_{}".format(chan)] = self.psu.get_curr(chan)
+                            except socket.timeout as e:
+                                self.log.warn(f"TimeoutError: {e}")
+                                self.log.info("Attempting to reconnect")
+                                self.psu = None
+                                break
+
+                    if self.psu:  # Check to make sure there weren't any timeout errors so we don't publish incomplete data and make block errors.
+                        # self.log.info(str(data))
+                        self.agent.publish_to_feed('psu_output', data)
+
+                        # Allow this process to be queried to return current data
+                        session.data = data
+
+                    else:
+                        if not self._initialize_module(session, params):
+                            time.sleep(5)  # wait 5 sec before trying to re-initialize
```

Comment:
This reconnect logic looks good and can be kept. Just to offer a suggestion for a less nested approach (since this is already a heavily nested block) I would have done something like this pseudo-code:

```
while monitor:
    if not psu:
        # reconnect to PSU
        # check connection, if not connected continue

    try:
        # take data
    except:
        psu = None
        continue

    # publish data
```

Essentially ensure the connection is established right off the bat, else continue to the next loop iteration where we try to re-establish connection. After which you can assume the connection is good when you try to take data, if that fails, continue to the 're-establish connection/take data' in the next loop, then publish data.

Reply:
I slightly modified the structure to the following. The if psu check before publishing is needed before publishing data to prevent block errors.
```
while monitor:
    if not acquired:
        # warn

    elif not psu:
        # reconnect to PSU
        # check connection, if not connected continue
    
    else:
        for chan in channels:
            try:
                # take data
            except:
                psu = None
                break
        
        if psu:
            # publish data
```

Reply:
I found a better solution using `continue ` to reduce nesting. I moved the sleep command to the beginning since using continue means the rest of while loop is skipped and placing 4 identical sleep command for the same purpose does not seem ideal. Psuedo-code of the new structure is below for ease of understanding:
```
while monitor:
    # sleep
    if not acquired:
        # warn
        continue

    if not psu:
        # reconnect to PSU
        continue
    
    try:
        for chan in channels:
            # take data
    except:
        psu = None
        continue

    # publish data
```

Reply:
This looks great, thanks for the updates!